example/poisson_gmg/poisson_equation.hpp

```diff
@@ -0,0 +1,218 @@
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#ifndef EXAMPLE_POISSON_GMG_POISSON_EQUATION_HPP_
+#define EXAMPLE_POISSON_GMG_POISSON_EQUATION_HPP_
+
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include <kokkos_abstraction.hpp>
+#include <parthenon/package.hpp>
+
+#include "poisson_package.hpp"
+
+namespace poisson_package {
+
+// This class implement methods for calculating A.x = y and returning the diagonal of A,
+// where A is the the matrix representing the discretized Poisson equation on the grid.
+// Here we implement the Laplace operator in terms of a flux divergence to (potentially)
+// consistently deal with coarse fine boundaries on the grid. Only the routines Ax and
+// SetDiagonal need to be defined for interfacing this with solvers. The other methods
+// are internal, but can't be marked private or protected because they launch kernels
+// on device.
+class PoissonEquation {
+ public:
+  bool do_flux_cor = false;
+
+  // Add tasks to calculate the result of the matrix A (which is implicitly defined by
+  // this class) being applied to x_t and store it in field out_t
+  template <class x_t, class out_t, class TL_t>
+  parthenon::TaskID Ax(TL_t &tl, parthenon::TaskID depends_on,
+                       std::shared_ptr<parthenon::MeshData<Real>> &md) {
+    auto flux_res = tl.AddTask(depends_on, CalculateFluxes<x_t>, md);
+    if (do_flux_cor && !(md->grid.type == parthenon::GridType::two_level_composite)) {
+      auto start_flxcor =
+          tl.AddTask(flux_res, parthenon::StartReceiveFluxCorrections, md);
+      auto send_flxcor = tl.AddTask(flux_res, parthenon::LoadAndSendFluxCorrections, md);
+      auto recv_flxcor = tl.AddTask(send_flxcor, parthenon::ReceiveFluxCorrections, md);
+      flux_res = tl.AddTask(recv_flxcor, parthenon::SetFluxCorrections, md);
+    }
+    return tl.AddTask(flux_res, FluxMultiplyMatrix<x_t, out_t>, md);
+  }
+
+  // Calculate an approximation to the diagonal of the matrix A and store it in diag_t.
+  // For a uniform grid or when flux correction is ignored, this diagonal calculation
+  // is exact. Exactness is (probably) not required since it is just used in Jacobi
+  // iterations.
```

Comment:
Also, this is a me problem, but I always have trouble unwrapping the shape of the mesh into matrix form... Not on you to document that though.

Reply:
Yeah, I think this is best left to the method documentation.